presto-main/src/main/java/com/facebook/presto/cost/CostCalculator.java

```diff
@@ -0,0 +1,228 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.cost;
+
+import com.facebook.presto.metadata.InternalNodeManager;
+import com.facebook.presto.spi.statistics.Estimate;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.EnforceSingleRowNode;
+import com.facebook.presto.sql.planner.plan.ExchangeNode;
+import com.facebook.presto.sql.planner.plan.FilterNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.LimitNode;
+import com.facebook.presto.sql.planner.plan.OutputNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanVisitor;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.planner.plan.SemiJoinNode;
+import com.facebook.presto.sql.planner.plan.TableScanNode;
+import com.facebook.presto.sql.planner.plan.ValuesNode;
+
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.function.Function;
+
+import static com.facebook.presto.cost.PlanNodeCostEstimate.UNKNOWN_COST;
+import static com.facebook.presto.cost.PlanNodeCostEstimate.ZERO_COST;
+import static com.facebook.presto.cost.PlanNodeCostEstimate.cpuCost;
+import static com.facebook.presto.spi.statistics.Estimate.zeroValue;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Simple implementation of CostCalculator.
+ */
+@ThreadSafe
+public class CostCalculator
+{
+    private final int numberOfNodes;
+
+    @Inject
+    public CostCalculator(InternalNodeManager nodeManager)
+    {
+        this(nodeManager.getAllNodes().getActiveNodes().size());
+    }
+
+    public CostCalculator(int numberOfNodes)
+    {
+        this.numberOfNodes = numberOfNodes;
+    }
+
+    public PlanNodeCostEstimate calculateCost(
```

Comment:
Imo it would be better to pass constrained `Lookup` which allows querying for self and children to `CostCalculator.calculateCost` and `StatsCalculator.calculateStats`.

IMO this is beneficial as it shows the approach that we want to treat statistics/cost as traits. 
And also allows for changes in implementation of stats/cost calcuator without changing interface.